src/Compilers/CSharp/Portable/Symbols/TypeWithAnnotations.cs

```diff
@@ -585,27 +592,43 @@ private static bool SkipNullableTransform(TypeSymbol type)
 
         public void AddNullableTransforms(ArrayBuilder<byte> transforms)
         {
-            var type = Type;
+            AddNullableTransforms(this, transforms);
+        }
 
-            if (!SkipNullableTransform(type))
+        private static void AddNullableTransforms(TypeWithAnnotations typeWithAnnotations, ArrayBuilder<byte> transforms)
+        {
+            while (true)
             {
-                byte flag;
-                if (NullableAnnotation.IsOblivious() || type.IsValueType)
-                {
-                    flag = NullableAnnotationExtensions.ObliviousAttributeValue;
-                }
-                else if (NullableAnnotation.IsAnnotated())
+                var type = typeWithAnnotations.Type;
+
+                if (!SkipNullableTransform(type))
                 {
-                    flag = NullableAnnotationExtensions.AnnotatedAttributeValue;
+                    var annotation = typeWithAnnotations.NullableAnnotation;
+                    byte flag;
+                    if (annotation.IsOblivious() || type.IsValueType)
+                    {
+                        flag = NullableAnnotationExtensions.ObliviousAttributeValue;
+                    }
+                    else if (annotation.IsAnnotated())
+                    {
+                        flag = NullableAnnotationExtensions.AnnotatedAttributeValue;
+                    }
+                    else
+                    {
+                        flag = NullableAnnotationExtensions.NotAnnotatedAttributeValue;
+                    }
+                    transforms.Add(flag);
                 }
-                else
+
+                if (type.TypeKind != TypeKind.Array)
                 {
-                    flag = NullableAnnotationExtensions.NotAnnotatedAttributeValue;
+                    type.AddNullableTransforms(transforms);
+                    return;
                 }
-                transforms.Add(flag);
-            }
 
-            type.AddNullableTransforms(transforms);
+                // Avoid recursion to allow for deeply-nested arrays.
+                typeWithAnnotations = ((ArrayTypeSymbol)type).ElementTypeWithAnnotations;
```

Comment:
>typeWithAnnotations = ((ArrayTypeSymbol)type).ElementTypeWithAnnotations; [](start = 16, length = 73)

Should ```ApplyNullableTransforms``` handle array types specially too?

Reply:
`ApplyNullableTransforms` would require a stack since we need to apply the byte for the element before the byte to the array.

---
In reply to: [296420854](https://github.com/dotnet/roslyn/pull/36152#discussion_r296420854) [](ancestors = 296420854)